### PR TITLE
KWSys 2023-01-19 (be3c441e)

### PR DESCRIPTION
**Pull request by dzenanz**
Code extracted from:

    https://gitlab.kitware.com/utils/kwsys.git

at commit be3c441e46e091a7606565221bb5ae7c9a9b684f (master).

Upstream Shortlog
-----------------

Aaron Liu (1):
      dc6b6641 testSystemTools: apply root-check on AIX

Alexey Edelev (1):
      549d3d0b SystemTools: Fix type of GetLineFromStream

Ben Boeckel (1):
      4226d5e5 Status: detect and diagnose X11 symbol conflicts

Brad King (12):
      929b6c6c Glob: Revert "Optimize directory/symlink checks on Windows"
      4b552447 Directory: Replace FileData with methods accepting file index
      259cdb6a SystemTools: Update SetPermissions test case to avoid 0 perms on POSIX
      591c4efb Activate POSIX APIs even without compiler extensions
      45f20b83 Fix XL/XLClang warning suppression under CMP0089 NEW behavior
      257b1111 Configure with policies through CMake 3.22
      b72169e5 Process: Suppress clang -Wshorten-64-to-32 diagnostic on macOS
      30e10c87 SystemTools: Report with copy operation failures which path failed
      6c66ba9e clang-format.bash: Use generic clang-format attribute
      82ae3f28 clang-format.bash: update to clang-format-15
      a61d0ad6 Empty commit at end of history preceding clang-format-15 style transition
      3cb35bf3 CONTRIBUTING: Update documented clang-format version to 15

Christopher Chavez (1):
      5a33629f ProcessUNIX.c: fix select(), FD_SETSIZE on macOS

Clemens Wasser (11):
      43ce7a20 SystemTools: Factor out FileIsSymlinkWithAttr helper
      d078f9e6 Directory: Store FIND_DATA for files in Directory
      7573b0fd Directory: Add Is{Directory,Symlink} to FileData
      99c7831e Glob: Optimize directory/symlink checks on Windows
      d4c5ed92 Glob: Optimize directory/symlink checks on Windows
      ab7c8f6f SystemTools: Factor out RemoveTrailingSlashes
      ca9646bd SystemTools: Optimize FileIsExecutable on Windows
      1c156fd1 SystemTools: Better GetCasePathName caching
      9e48ae84 SystemTools: Readd direct GetCasePathName caching
      0cac8ee5 Directory: Use Win32 find file functions
      550b5734 SystemTools: Use unordered_map for path caches

Dimitris Apostolou (1):
      b387faf8 Process: Fix typos

Jessica Clarke (1):
      ebfb5cdb SystemInformation: Change GetRealAddress to return a size_t

Joker-Van (1):
      dfec20c9 SystemInformation: Add Zhaoxin x86 processor kh40000

Kitware Robot (1):
      f685d817 Revise C++ coding style using clang-format-15

Markus87 (1):
      5f2dcc13 SystemTools: Fix FilesDiffer 32-bit signed integer overflow on Windows

Michael Hirsch (1):
      f0223ad1 SystemInformation: correct function name spelling

Saleem Abdulrasool (1):
      0969597d SystemTools: add a clonefile optimization on macOS

Sean McBride (1):
      d6c6fd82 testDirectory: Rename functions to fix -Wreserved-identifier warnings

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

**Comment by dzenanz** [comment]
@bradking was there some change in KWSys which would require update of [itkDirectoryTest](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/test/itkDirectoryTest.cxx)? It now fails: https://open.cdash.org/test/932890211.

**Review by tbirdso — COMMENTED** [review]
Looks good at a glance, would be good to have a review from someone more familiar with KWSys.

**Comment by bradking** [comment]
The `itkDirectoryTest` case in question is indexing out of bounds, which was never documented as supported.  Since [KWSys MR 243](https://gitlab.kitware.com/utils/kwsys/-/merge_requests/243) the `Directory::GetFile` method does not do an upper-bound check, and since [KWSys MR 247](https://gitlab.kitware.com/utils/kwsys/-/merge_requests/247), the other `Directory::Get*` methods don't either.

File: Modules/Core/Common/test/itkDirectoryTest.cxx [hunk 1]
```diff
@@ -45,14 +45,8 @@ itkDirectoryTest(int argc, char * argv[])
...
-  // This should fail
-  auto fileOutOfRange = static_cast<unsigned int>(directory->GetNumberOfFiles());
-  if (directory->GetFile(fileOutOfRange))
-  {
-    std::cerr << "directory->GetFile(directory->GetNumberOfFiles())"
-              << " should have failed." << std::endl;
-    return EXIT_FAILURE;
-  }
+  // Since 78dcc270826fe87fac5bf46d65ca6f02d9f47b4b the next line causes a crash:
+  // auto fileOutOfRange = static_cast<unsigned int>(directory->GetNumberOfFiles());
```
Comment (bradking) [inline Modules/Core/Common/test/itkDirectoryTest.cxx:49]: There is no reason to support out-of-bounds indexing, so I suggest just revising this commit to remove that test case.

**Pull request merged**